vsintegration/src/FSharp.Editor/Navigation/GoToDefinition.fs

```diff
@@ -726,4 +728,216 @@ type internal FSharpNavigation
+        match s with
+        | "E" -> Event
+        | "P" -> Property
+        | "CTOR" -> Constructor // That one is "artificial one", so we distinguish constructors.
+        | "M" -> Method
+        | _ -> Other
+
+type internal SymbolPath = { EntityPath: string list; MemberOrValName: string; GenericParameters: int }
+
+[<RequireQualifiedAccess>]
+type internal DocCommentId =
+    | Member of SymbolPath * SymbolMemberType: SymbolMemberType
+    | Field of SymbolPath
+    | Type of EntityPath: string list
+    | None
+
+type FSharpNavigableLocation(statusBar: StatusBar, metadataAsSource: FSharpMetadataAsSourceService, symbolRange: range, project: Project) =
+    interface IFSharpNavigableLocation with
+        member _.NavigateToAsync(_options: FSharpNavigationOptions2, cancellationToken: CancellationToken) : Task<bool> =
+            asyncMaybe {
+                let targetPath = symbolRange.FileName
+                let! targetDoc = project.Solution.TryGetDocumentFromFSharpRange (symbolRange, project.Id)
+                let! targetSource = targetDoc.GetTextAsync(cancellationToken)
+                let gtd = GoToDefinition(metadataAsSource)
+
+                let (|Signature|Implementation|) filepath =
+                    if isSignatureFile filepath then Signature else Implementation
+                
+                match targetPath with
+                | Signature  ->
+                    return! gtd.NavigateToSymbolDefinitionAsync(targetDoc, targetSource, symbolRange, statusBar, cancellationToken)
+                | Implementation ->
+                    return! gtd.NavigateToSymbolDeclarationAsync(targetDoc, targetSource, symbolRange, statusBar, cancellationToken)
+            }
+            |> Async.map (fun a -> a.IsSome)
+            |> RoslynHelpers.StartAsyncAsTask cancellationToken
+
+[<Export(typeof<IFSharpCrossLanguageSymbolNavigationService>)>]
+[<Export(typeof<FSharpCrossLanguageSymbolNavigationService>)>]
+type FSharpCrossLanguageSymbolNavigationService()  =
+    let componentModel = Package.GetGlobalService(typeof<ComponentModelHost.SComponentModel>) :?> ComponentModelHost.IComponentModel
+    let workspace = componentModel.GetService<VisualStudioWorkspace>()
+    let statusBar = StatusBar(ServiceProvider.GlobalProvider.GetService<SVsStatusbar,IVsStatusbar>())
+    let metadataAsSource = componentModel.DefaultExportProvider.GetExport<FSharpMetadataAsSourceService>().Value
+
+    // So, the groups are following:
+    //   1 - type (see below).
+    //   2 - Path - a dotted path to a symbol.
+    //   3 - parameters, opetional, only for methods and properties.
+    //   4 - return type, optional, only for methods.
+    let docCommentIdRx = Regex(@"^(\w):([\w\d#`.]+)(\(.+\))?(?:~([\w\d.]+))?$", RegexOptions.Compiled)
+
+    // Parse generic args out of the function name
+    let fnGenericArgsRx = Regex(@"^(.+)``(\d+)$", RegexOptions.Compiled)
+
+    let docCommentIdToPath (docId:string) =
```

Comment:
This function is big enough on it's own.
If it would exists outside of the type, it could be also separately unit-testable. (which I think it deserves, even though we do not do many unit tests, here it kind of makes sense)

Reply:
Sure, although it's a throwaway function.